src/test/clojure/xtdb/trie_test.clj

```diff
@@ -173,23 +173,89 @@
-                                          (->mock-merge-plan-page 2 false)])
-                      (map :page)))
-        "earlier pages don't need to be taken if metadata doesn't match")
-
-  (t/is (= [0 1 2] (->> (trie/->merge-task [(->mock-merge-plan-page 0 true)
-                                            (->mock-merge-plan-page 1 false)
-                                            (->mock-merge-plan-page 2 false)])
-                        (map :page)))
-        "all pages pages need to be taken if the earliest matches"))
+(def ^:private inst->micros (comp time/instant->micros time/->instant))
 
+(deftest earilier-recency-buckets-can-effect-splitting-in-later-buckets-4097
+  (let [[system-from1 system-from2] (mapv inst->micros [#inst "2020-01-01" #inst "2020-01-03"])
+        [valid-from1 valid-from2] (mapv inst->micros [#inst "2020-01-01" #inst "2020-01-03"])
+        [valid-to1 valid-to2] (mapv inst->micros [#inst "2021-01-05" #inst "2020-01-04"])]
+    (t/is (= [1 0] (->> (trie/->merge-task
+                         [(->mock-merge-plan-page 0 false valid-from1 valid-to1 system-from1)
+                          (->mock-merge-plan-page 1 true  valid-from2 valid-to2 system-from2)])
+                        (map :page)))
+          "earlier pages that contain data with later system time need to be taken")))
 
-(def ^:private inst->micros (comp time/instant->micros time/->instant))
+(deftest test-to-merge-task
+  (let [insts (tu/->instants :day)
+        [sf1 sf2 sf3] (->> insts (take 3) (map inst->micros))
+        [vf1 vf2 vf3] (->> insts (take 3) (map inst->micros))
+        [vt1 vt2 vt3] (->> insts (drop 1) (take 3) (map inst->micros))
+        ->pages #(into #{} (map :page) %)]
+
+    ;; First page comes before in system time
+    ;; Third page doesn't overlap in valid time
+    (t/is (= #{1} (->> (trie/->merge-task [(->mock-merge-plan-page 0 false vf1 vt1 sf1 Long/MAX_VALUE sf1)
+                                           (->mock-merge-plan-page 1 true vf2 vt2 sf2 Long/MAX_VALUE sf2)
+                                           (->mock-merge-plan-page 2 false vf3 vt3 sf3 Long/MAX_VALUE sf3)]
+                                          (tu/->min-max-page-bounds vf2 Long/MAX_VALUE sf2 Long/MAX_VALUE))
+                       ->pages))
+          "No need to take extra pages!")
+
+    (t/is (= #{1 2} (->> (trie/->merge-task [(->mock-merge-plan-page 0 false vf1 (inc vt1) sf1 Long/MAX_VALUE sf1)
+                                             (->mock-merge-plan-page 1 true vf2 (inc vt2) sf2 Long/MAX_VALUE sf2)
+                                             (->mock-merge-plan-page 2 false vf3 (inc vt3) sf3 Long/MAX_VALUE sf3)]
+                                            (tu/->min-max-page-bounds vf2 Long/MAX_VALUE sf2 Long/MAX_VALUE))
+                         ->pages))
+          "We make the valid time windows slightly larger so it overlaps. Third page needs to get taken.")
```

Comment:
let's express these messages in terms of allen intervals

Reply:
I added some comments in that test. I wasn't able to purely write it in terms Allen intervals.